jcache/commons/src/main/java/org/infinispan/jcache/AbstractJCacheNotifier.java

```diff
@@ -76,44 +85,129 @@ public void removeListener(CacheEntryListenerConfiguration<K, V> reg,
          jcache.removeListener(listenerAdapter);
    }
 
+   public void addSyncNotificationLatch(Cache<K, V> cache, K key, V value, CountDownLatch latch) {
+      EventSource<K, V> eventSourceKey = new EventSource<K, V>(cache, key, value);
+
+      Queue<CountDownLatch> latches = new ConcurrentLinkedQueue<>();
+      Queue<CountDownLatch> prev = latchesByEventSource.putIfAbsent(eventSourceKey, latches);
+      if (prev != null) {
+         latches = prev;
+      }
+      latches.add(latch);
+   }
+
+   public void removeSyncNotificationLatch(Cache<K, V> cache, K key, V value, CountDownLatch latch) {
+      EventSource<K, V> eventSourceKey = new EventSource<K, V>(cache, key, value);
+
+      Queue<CountDownLatch> latches = latchesByEventSource.get(eventSourceKey);
+
+      if (latches == null) {
+         return;
+      }
+      latches.remove(latch);
+      if (latches.isEmpty()) {
+         latchesByEventSource.remove(eventSourceKey, EMPTY_QUEUE);
```

Comment:
> This doesn't seem like it wouldn't remove anything.

AFAICT from the prev PR, it will remove the event from the map when the queue of latches is empty.

> Also we would need to use the compute method to properly do this in an atomic fashion.

Maybe that could eliminate the need of the ~~UnsupportedOperationQueue~~ EmptyQueue :+1: